Plugins/NativeCapture/Source/NativeCapture.cpp

```diff
@@ -0,0 +1,81 @@
+#include "NativeCapture.h"
+
+#include <Babylon/JsRuntime.h>
+#include <GraphicsImpl.h>
+
+#include <vector>
+
+namespace Babylon::Plugins::Internal
+{
+    class NativeCapture : public Napi::ObjectWrap<NativeCapture>
+    {
+    public:
+        static constexpr auto JS_CLASS_NAME = "NativeCapture";
+
+        static void Initialize(Napi::Env env)
+        {
+            Napi::HandleScope scope{env};
+
+            Napi::Function func = NativeCapture::DefineClass(
+                env,
+                JS_CLASS_NAME,
+                {
+                    NativeCapture::InstanceMethod("addOnCaptureCallback", &NativeCapture::AddOnCaptureCallback),
+                    NativeCapture::InstanceMethod("dispose", &NativeCapture::Dispose),
+                });
+
+            env.Global().Set(JS_CLASS_NAME, func);
+        }
+
+        NativeCapture(const Napi::CallbackInfo& info)
+            : Napi::ObjectWrap<NativeCapture>{info}
+            , m_runtime{JsRuntime::GetFromJavaScript(info.Env())}
+            , m_graphicsImpl(Graphics::Impl::GetFromJavaScript(info.Env()))
+            , m_ticket{m_graphicsImpl.AddCaptureCallback([this](auto& data) { CaptureDataReceived(data); })}
+        {
+            m_graphicsImpl.StartCapture();
+        }
+
+    private:
+        void AddOnCaptureCallback(const Napi::CallbackInfo& info)
+        {
+            auto listener = info[0].As<Napi::Function>();
+            m_callbacks.push_back(Napi::Persistent(listener));
+        }
+
+        void CaptureDataReceived(const BgfxCallback::CaptureData& data)
+        {
+            std::vector<uint8_t> bytes{};
+            bytes.resize(data.DataSize);
+            std::memcpy(bytes.data(), data.Data, data.DataSize);
+            m_runtime.Dispatch([this, data = data, bytes = std::move(bytes)](Napi::Env env) mutable {
+                data.Data = bytes.data();
+
+                auto external = Napi::External<BgfxCallback::CaptureData>::New(env, &data);
+                for (const auto& callback : m_callbacks)
+                {
+                    callback.Call({external});
+                }
+            });
+        }
+
+        void Dispose(const Napi::CallbackInfo&)
```

Comment:
What happens if dispose isn't called? Should we call dispose in the destructor automatically?

Reply:
I guess we have to. This really shouldn't be allowed -- the purpose of having dispose is to ensure that a capture's lifetime is controlled by program logic as opposed to the whims of the garbage collector -- but we may not have a choice because the JS runtime can theoretically be torn down while the graphics object survives, so I suppose we need a fallback in the destructor to avoid potentially leaving graphics in a bad state.